noir-projects/noir-contracts/contracts/contract_class_registerer_contract/src/main.nr

```diff
@@ -217,13 +219,19 @@ pub contract ContractClassRegisterer {
         // cannot prove non-registration. Therefore, it is possible that a malicious oracle might prevent sequencers
         // from including transactions with calls to certain badly-broadcasted contracts.
         // TODO(#8978): review correctness
+        let log_to_emit: [Field; CONTRACT_CLASS_LOG_SIZE_IN_FIELDS] =
+            array_concat(log, [0; CONTRACT_CLASS_LOG_SIZE_IN_FIELDS - N]);
+        // TODO(MW): use chunked poseidon if cheaper? Use variable poseidon with length of non-0 fields?
         let log_hash =
-            unsafe { emit_contract_class_unencrypted_log_private(contract_address, log, counter) };
+            std::hash::poseidon2::Poseidon2::hash(log_to_emit, CONTRACT_CLASS_LOG_SIZE_IN_FIELDS);
+        // Safety: TODO(MW): constraining below length in the base rollup. Is this good/safe?
+        let length = unsafe { find_index_hint_from_end(log_to_emit, |elem| !is_empty(elem)) };
+        // Safety: the below only exists to broadcast the raw log, so we can provide it to the base rollup later to be constrained.
```

Comment:
Two Qs here:
- I'm just hashing the entire (padded) log here because it's easy and clean. I could hash just the N input fields. This is a bit more expensive in the circuits (not by much, I'd use the chunk variable hash) but might be a lot faster elsewhere. I already have code to hash N of M fields, and check the remaining M - N fields are all zero.
- I'm carrying the log length (N) throughout the kernels to track gas since we are only adding N things to the blob. I'm not sure this is even required atm because logs are set to cost 0? I'm also not constraining this length until the base rollup when we verify the hash. I'm not sure if this is ok.

Reply:
I'd potentially go with the unpadded approach, since our recent discussions (via that hackmd) suggest that eventually (probably in a separate PR), we don't actually need to hash in this circuit at all.

We probably do need to track the length, because ultimately these logs should cost a relative fortune. And eventually (as per prev paragraph) only the Base will need to constrain this length (because we'll remove constrained hashing of this log from private-land).

Reply:
> because we'll remove constrained hashing of this log from private-land

Shall I just do this now, and make the oracle hash again? I don't remember now - but wasn't there some reason we weren't happy with the original oracle hash even though we verified the hash later?

Reply:
Sorry for the delay; I need to sort out my notification settings.
> Shall I just do this now, and make the oracle hash again?

Can do! It can either be an oracle, or just make the function `unconstrained`? But you messaged 14 hours ago, so maybe you've already chosen an approach, which I'll find out when I read further down this page :)

Reply:
> wasn't there some reason we weren't happy with the original oracle hash even though we verified the hash later?

I can't remember 😬

Reply:
Just FYI Alvaro found the unpadded hash is insecure unless we also constrain the length in the registeter (which just adds gates. Fixed len hash is about the same gates). From @sirasistant:

> I think that opens the attack vector:
> - Private chooses a shorter length
> - Private only hashes the shorter length
> - Base rollup receives the shorter preimage
> - Hashes match
> - We publish incomplete bytecode

I willl revert to a fixed length hash